crypto/src/mls/credential/trust_anchor.rs

```diff
@@ -0,0 +1,1647 @@
+        if domain_name != self.domain_name {
+            return Err(CryptoError::DomainNamesDontMatch);
+        }
+
+        // verify the whole chain
+        let root_cert = certificate_chain
+            .iter()
+            .map(Ok)
+            .reduce(
+                |child, parent| -> Result<&Certificate, openmls_traits::types::CryptoError> {
+                    let child = child?;
+                    let parent = parent?;
+                    child.is_valid()?;
+                    child.is_signed_by(backend.crypto(), parent)?;
+                    Ok(parent)
+                },
+            )
+            .unwrap()
+            .map_err(MlsError::from)?;
+        // ensure that the root is also valid
+        root_cert.is_valid().map_err(MlsError::from)?;
+
+        check_root_in_trust_store(root_cert)?;
+
+        let encoded_chain = certificate_chain
+            .into_iter()
+            .map(|cert| cert.to_der())
+            .collect::<Result<_, x509_cert::der::Error>>()?;
+        Ok(encoded_chain)
+    }
+}
+
+/// Checks the root cert against the trust store. In wasm maybe use webpki-roots (https://github.com/rustls/webpki-roots) crate
+fn check_root_in_trust_store(_root: &Certificate) -> CryptoResult<()> {
+    // TODO: verify root certificate is valid in the device's trust store
+    Ok(())
+}
+
+impl MlsConversation {
+    /// Validates the certificate chain
+    pub(crate) fn validate_received_trust_anchors(
+        old_group_context: &GroupContext,
+        commit_group_context: &GroupContext,
+        backend: &MlsCryptoProvider,
+    ) -> CryptoResult<()> {
+        if let Some(anchors) = commit_group_context.extensions().per_domain_trust_anchors() {
+            // find new anchors
+            anchors
+                .iter()
+                .filter(|anchor| {
+                    old_group_context.extensions().per_domain_trust_anchors().is_none()
+                        || old_group_context
+                            .extensions()
+                            .per_domain_trust_anchors()
+                            .is_some_and(|old_anchors| {
+                                !old_anchors
+                                    .iter()
+                                    .any(|old_anchor| old_anchor.certificate_chain() == anchor.certificate_chain())
+                            })
+                })
+                .try_for_each(|anchor| -> CryptoResult<()> {
+                    let anchor = PerDomainTrustAnchor::try_from(anchor)?;
+                    // the domain will obviously be in the context and will be already applied by
+                    // other client, so the domain uniquennes should not be validated here
+                    anchor.validate(backend, None)?;
+                    Ok(())
+                })?;
+        }
+        Ok(())
+    }
+
+    /// see [MlsCentral::update_trust_anchors_from_conversation]
+    #[cfg_attr(test, crate::durable)]
+    pub(crate) async fn update_trust_anchors(
```

Comment:
You should really extract the validation part of this method into a dedicated method. This way we can reuse it if/when we allow to do the same in a GCE proposal